xds/src/main/java/io/grpc/xds/ClientXdsClient.java

```diff
@@ -257,12 +263,58 @@ protected void handleLdsResponse(String versionInfo, List<Any> resources, String
     }
   }
 
+  @VisibleForTesting
+  @Nullable // Returns null if the filter is optional but not supported.
+  static StructOrError<FilterConfig> parseHttpFilter(
+      io.envoyproxy.envoy.extensions.filters.network.http_connection_manager.v3.HttpFilter
+          httpFilter) {
+    String filterName = httpFilter.getName();
+    if (!httpFilter.hasTypedConfig()) {
+      if (httpFilter.getIsOptional()) {
+        return null;
+      } else {
+        return StructOrError.fromError(
+            "HttpFilter [" + filterName + "] is not optional and has no typed config");
+      }
+    }
+    Any anyConfig = httpFilter.getTypedConfig();
+    Message rawConfig = anyConfig;
+    String typeUrl = anyConfig.getTypeUrl();
+    if (anyConfig.getTypeUrl().equals(TYPE_URL_TYPED_STRUCT)) {
```

Comment:
>Can we somewhat reuse the code for populating the typed struct, instead of repeating it for parseHttpFilter() and parseOverrideFilterConfigs()?

It's only a few lines of duplicate code, it's not so easy to reuse the code for java because it needs to populate pair-value [typeUrl, rawConfig]. In addition, `TYPE_URL_TYPED_STRUCT` support is currently dead code.

> Still, related to my previous comment on having parseFilterConfig(Message rawProtoMessage) and parseFilterConfigOverride(Message rawProtoMessage) instead of one unified method with boolean. 

In general `parseFilterConfig` and `parseFilterConfigOverride` can have completely different implementation, because the override config can be of a completely different type. It's only coincidence that the type of top-level config and override config are the same for Fault Filter.

Reply:
> It's only a few lines of duplicate code, it's not so easy to reuse the code for java because it needs to populate pair-value [typeUrl, rawConfig]. In addition, `TYPE_URL_TYPED_STRUCT` support is currently dead code.

Hmm... That's still a fair amount of code repeat. I think you could write a helper method that parses a single `Any` and it could be called for parsing both top-level config or override config. And, since it's a private helper method, it's absolutely ok to use a boolean argument to indicate whether calls `parseFilterConfig(...)` or `parseFilterConfigOverride(...)`.

You helper method can be pure. That is, do not consider whatever the typeUrl the `Any` argument would be, just branch to cover all cases.

Reply:
>  I think you could write a helper method that parses a single Any and it could be called for parsing both top-level config or override config. And, since it's a private helper method, it's absolutely ok to use a boolean argument to indicate whether calls parseFilterConfig(...) or parseFilterConfigOverride(...).

That can be achieved. Done.